README.md
# castv2-device-monitor

Monitor the state of a castv2 (a.k.a. Chromecast) device.

## Usage

    const DeviceMonitor = require('castv2-device-monitor).DeviceMonitor

    // Monitor the chromecast with the friendly name Livingroom
    let dm = new DeviceMonitor('Livingroom')
    dm.on('powerState', powerState => console.log('powerState', powerState))
    dm.on('playState', playState => console.log('playState', playState))
    dm.on('application', application => console.log('application', application))
    dm.on('media', media => console.log('media', media))

A DeviceMonitor monitors a single Chromecast device and emits events whenever something changes with it.
It can emit the following events:

### event 'powerState'

The powerState event has a single parameter with the possible values 'on' and 'off'.
A DeviceMonitor will always emit a powerState event soon after it is created.

As far as I know, a Chromecast device does not expose any real power indicator.
The powerState is an attempt to emulate a power indicator. powerState will always
be on when the Chromecast is playing something, and it will be set to
off after the Chromecast has been idle for 60 seconds.
To use another idle timeout, supply the timeout in milliseconds
as the third parameter to the DeviceMonitor contructor.

The powerState can be used to control power to
an amplifier or TV that the Chromecast is connected to.

### event 'playState'

The playState event has a single parameter with the possible values 'play' and 'pause'.

### event 'application'

The application event has a single parameter that is the friendly name of
the application that is currently controlling the Chromecast.

### event 'media'

The media event has a single parameter that contains an object with properties
'artist' and 'title'. These describe the currently playing song on the Chromecast.

### event 'volume'
The volume event has a single parameter that contains the current volume as
a number between 0-1, where 0 is muted and 1 is the maximum volume.

# Controlling device

The DeviceMonitor also offers a limited ability to control playback on the device:

    dm.pauseDevice()
    dm.playDevice()
    dm.volumeUp()
    dm.volumeDown()

Additionally, dm.stopDevice() can be used to stop the session
from the application to the device.
It usually means that you have to
select the cast-device again from the application.

# Debugging

This module includes a binary castv2-device-monitor for testing purposes.

Usage:

    ./node_modules/.bin/castv2-device-monitor Livingroom

    DEBUG=* ./node_modules/.bin/castv2-device-monitor Livingroom eth0 5000


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.3.2] - 2018-08-20

### Added

- Expose additional metadata. Patch by nickcmaynard.

## [1.3.1] - 2018-08-11

### Changed

- Set powerState to off when the media channel is closed. Patch by nickcmaynard.


monitor-chromecast.js
var Client = require('castv2').Client
var DefaultMediaReceiver = require('castv2').DefaultMediaReceiver
var mdns = require('mdns')

var browser = mdns.createBrowser(mdns.tcp('googlecast'))

// http://bagaar.be/index.php/blog/on-chromecasts-and-slack

var requestId = 1

browser.on('serviceUp', service => {
    if (service.txtRecord.fn === 'Garage') {
        browser.stop()
        var client = new Client()
        client.connect(service.addresses[0], () => {
            const connection = client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.tp.connection',
                'JSON',
            )
            const receiver = client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.receiver',
                'JSON',
            )
            const heartbeat = client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.heartbeat',
                'JSON',
            )
            connection.send({ type: 'CONNECT' })
            setInterval(() => {
                heartbeat.send({ type: 'PING' })
            }, 5000)
            receiver.send({ type: 'GET_STATUS', requestId: requestId++ })
            receiver.on('message', data => {
                if (!data.status.applications) return
                let appID = data.status.applications[0].transportId
                let mediaConnection = client.createChannel(
                    'client-17558',
                    appID,
                    'urn:x-cast:com.google.cast.tp.connection',
                    'JSON',
                )
                let media = client.createChannel(
                    'client-17558',
                    appID,
                    'urn:x-cast:com.google.cast.media',
                    'JSON',
                )
                mediaConnection.send({ type: 'CONNECT' })
                media.send({ type: 'GET_STATUS', requestId: requestId++ })
                media.on('message', songInfo => parseData(songInfo))
            })
        })
    }
})

function parseData(songInfo) {
    console.log('songInfo', JSON.stringify(songInfo))
}

browser.start()

function ondeviceup(host) {
    var client = new Client()

    client.connect(host, function() {
        // create various namespace handlers
        var connection = client.createChannel(
            'sender-0',
            'receiver-0',
            'urn:x-cast:com.google.cast.tp.connection',
            'JSON',
        )
        var heartbeat = client.createChannel(
            'sender-0',
            'receiver-0',
            'urn:x-cast:com.google.cast.tp.heartbeat',
            'JSON',
        )
        var receiver = client.createChannel(
            'sender-0',
            'receiver-0',
            'urn:x-cast:com.google.cast.receiver',
            'JSON',
        )
        var media = client.createChannel(
            'sender-0',
            'receiver-0',
            'urn:x-cast:com.google.cast.media',
            'JSON',
        )

        // establish virtual connection to the receiver
        connection.send({ type: 'CONNECT' })

        // start heartbeating
        setInterval(function() {
            heartbeat.send({ type: 'PING' })
        }, 5000)

        receiver.send({ type: 'GET_STATUS', requestId: 1 })
        media.send({ type: 'GET_STATUS', requestId: 2 })

        // display receiver status updates
        receiver.on('message', function(data, broadcast) {
            console.log(JSON.stringify(data))
        })

        media.on('message', function(data, broadcast) {
            console.log(JSON.stringify(data))
        })
    })
}


monitor-chromecast.ts
import { Client } from 'castv2'
import * as mdns from 'mdns'
import { EventEmitter } from 'events'
import * as dbg from 'debug'

const debug = dbg('cm')

let requestId = 1
export type PowerState = 'on' | 'off'
export type PlayState = 'play' | 'pause'
export type Media = {
    artist: string
    title: string
}

export class ChromecastMonitor extends EventEmitter {
    public powerState: PowerState | undefined = undefined
    public playState: PlayState = 'pause'
    public application: string | undefined = undefined

    public currentMedia: Media = {
        artist: 'none',
        title: 'none',
    }

    private found = false

    private serviceName: string
    private clientConnection: ClientConnection
    private idleTimer: NodeJS.Timer

    constructor(
        private friendlyName: string,
        private networkInterface?: string,
        private idleTimeout: number = 60000,
    ) {
        super()

        setTimeout(() => {
            if (!this.powerState) {
                this.setPowerState('off')
            }
        }, 2000)

        let browser = mdns.createBrowser(mdns.tcp('googlecast'))

        // http://bagaar.be/index.php/blog/on-chromecasts-and-slack

        let requestId = 1

        browser.on('serviceUp', service => {
            if (
                this.networkInterface &&
                this.networkInterface !== service.networkInterface
            ) {
                // Ignore wrong interface
            } else {
                if (service.txtRecord.fn === this.friendlyName) {
                    debug('serviceUp')
                    if (this.clientConnection) {
                        debug('Destroying clientConnection to replace with new')
                        this.clientConnection.close()
                    }
                    this.clientConnection = new ClientConnection(
                        service.addresses[0],
                        this,
                    )
                }
            }
        })

        browser.on('serviceDown', service => {
            if (
                this.networkInterface &&
                this.networkInterface !== service.networkInterface
            ) {
                // Ignore wrong interface
            } else {
                if (service.name === this.serviceName) {
                    debug('serviceDown')
                    this.setPowerState('off')
                    if (this.clientConnection) {
                        debug('No clientConnection to destroy')
                    } else {
                        this.clientConnection.close()
                    }
                }
            }
        })
        browser.start()
    }

    setPowerState(powerState: PowerState) {
        if (powerState !== this.powerState) {
            this.powerState = powerState
            this.emit('powerState', this.powerState)
        }

        if (this.powerState === 'on') {
            this.clearIdleTimer()
        }
    }

    setPlayState(playerState) {
        let playState: PlayState = 'play'

        if (
            playerState === 'PAUSED' ||
            playerState === 'IDLE' ||
            playerState === 'BUFFERING'
        ) {
            playState = 'pause'
            this.setIdleTimer()
        }

        if (playState === 'play') {
            this.setPowerState('on')
        }

        if (this.playState !== playState) {
            this.playState = playState
            this.emit('playState', this.playState)
        }
    }

    setMedia(media: Media) {
        if (
            media.artist !== this.currentMedia.artist ||
            media.title !== this.currentMedia.title
        ) {
            this.currentMedia = media
            this.emit('media', this.currentMedia)
        }
    }

    setApplication(application: string) {
        this.setPowerState('on')
        if (application !== this.application) {
            this.application = application
            this.emit('application', this.application)
        }
    }

    setIdleTimer() {
        this.clearIdleTimer()
        this.idleTimer = setTimeout(
            () => this.setPowerState('off'),
            this.idleTimeout,
        )
    }

    clearIdleTimer() {
        if (this.idleTimer) {
            clearTimeout(this.idleTimer)
            this.idleTimer = undefined
        }
    }
}

class ClientConnection {
    private client: Client
    private interval: NodeJS.Timer
    private active = false
    private reconnectTimer: NodeJS.Timer

    constructor(private address: string, private monitor: ChromecastMonitor) {
        this.connect()
    }

    connect() {
        this.active = true
        this.client = new Client()

        this.client.on('error', error => {
            console.log('error event', error)
            this.close()
            this.reconnectTimer = setTimeout(() => this.connect(), 5000)
        })

        this.client.connect(this.address, () => {
            debug('Connected')

            const connection = this.client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.tp.connection',
                'JSON',
            )
            const receiver = this.client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.receiver',
                'JSON',
            )
            const heartbeat = this.client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.heartbeat',
                'JSON',
            )

            connection.send({ type: 'CONNECT' })

            connection.on('message', data => {
                debug('connection message', data)
                if (data.type === 'CLOSE') {
                    // Never seen this happen
                    debug('Connection closed')
                    this.close()
                }
            })

            connection.on('close', () => debug('connection closed'))
            connection.on('disconnect', () => debug('connection disconnect'))

            this.interval = setInterval(() => {
                heartbeat.send({ type: 'PING' })
            }, 5000)

            receiver.send({
                type: 'GET_STATUS',
                requestId: requestId++,
            })

            receiver.on('message', data => {
                debug('receiver message', JSON.stringify(data))
                if (!data.status.applications) return
                this.monitor.setApplication(
                    data.status.applications[0].displayName,
                )
                let appID = data.status.applications[0].transportId
                new MediaConnection(this.client, appID, this.monitor)
            })
        })
    }

    close() {
        if (this.active) {
            this.active = false
            clearInterval(this.interval)
            this.client.close()
        } else {
            clearTimeout(this.reconnectTimer)
        }
    }
}

class MediaConnection {
    constructor(
        client: any,
        appId: string,
        private monitor: ChromecastMonitor,
    ) {
        let mediaConnection = client.createChannel(
            'client-17558',
            appId,
            'urn:x-cast:com.google.cast.tp.connection',
            'JSON',
        )

        let media = client.createChannel(
            'client-17558',
            appId,
            'urn:x-cast:com.google.cast.media',
            'JSON',
        )

        mediaConnection.send({ type: 'CONNECT' })

        mediaConnection.on('message', data => {
            debug('mediaConnection message', data)
            if (data.type === 'CLOSE') {
                // Happens when you press Stop from the cast-dialog in Chrome
                debug('Closing MediaConnection')
                mediaConnection.close()
                media.close()
            }
        })

        media.send({
            type: 'GET_STATUS',
            requestId: requestId++,
        })
        media.on('message', songInfo => this.parseData(songInfo))
    }

    parseData(songInfo) {
        debug('songInfo', songInfo.requestId)
        let status = songInfo.status[0]
        if (status) {
            this.monitor.setPlayState(status.playerState)
            if (status.media) {
                this.monitor.setMedia({
                    artist: status.media.metadata.artist,
                    title: status.media.metadata.title,
                })
            }
        } else {
            debug('No status')
        }
    }
}

let cm = new ChromecastMonitor('Garage', 'en0', 5000)
cm.on('powerState', powerState => console.log('powerState', powerState))
cm.on('playState', playState => console.log('playState', playState))
cm.on('application', application => console.log('application', application))
cm.on('media', media => console.log('media', media))


index.ts
const Client = require('castv2').Client
const mdns = require('mdns-js')

import { EventEmitter } from 'events'
import * as dbg from 'debug'
const debug = dbg('cm')

let requestId = 1
export type PowerState = 'on' | 'off'
export type PlayState = 'play' | 'pause'
export type Media = {
    artist: string
    title: string
    albumArtist?: string
    albumName?: string
    images?: Array<Image>
    trackNumber?: number
}
export type Image = {
    url: string
}
export type Volume = number;

export class DeviceMonitor extends EventEmitter {
    public powerState: PowerState | undefined = undefined
    public playState: PlayState = 'pause'
    public application: string | undefined = undefined

    public currentMedia: Media = {
        artist: 'none',
        title: 'none',
    }
    public currentVolume: Volume;

    private clientConnection: ClientConnection | undefined
    private idleTimer: NodeJS.Timer | undefined = undefined
    private clientIp: string

    constructor(
        private friendlyName: string,
        private idleTimeout: number = 60000,
    ) {
        super()

        setTimeout(() => {
            if (!this.powerState) {
                this.setPowerState('off')
            }
        }, 2000)

        let browser = mdns.createBrowser(mdns.tcp('googlecast'))
        browser.on('ready', function() {
            browser.discover()
        })

        // http://bagaar.be/index.php/blog/on-chromecasts-and-slack

        browser.on('update', (service: any) => {
            if (
                service.type[0].name === 'googlecast' &&
                service.txt &&
                service.txt.includes('fn=' + this.friendlyName)
            ) {
                let clientIp = service.addresses[0]
                if (clientIp !== this.clientIp) {
                    this.clientIp = clientIp
                    if (this.clientConnection) {
                        debug('Destroying clientConnection to replace with new')
                        this.clientConnection.close()
                    }
                    this.clientConnection = new ClientConnection(
                        this.clientIp,
                        this,
                    )
                }
                debug(service)
                debug('serviceUp')
            }
        })
    }

    setPowerState(powerState: PowerState) {
        if (powerState !== this.powerState) {
            this.powerState = powerState

            if (this.powerState === 'on') {
                this.clearIdleTimer()
            }

            this.emit('powerState', this.powerState)
        }
    }

    setPlayState(playerState: string) {
        let playState: PlayState = 'play'

        if (
            playerState === 'PAUSED' ||
            playerState === 'IDLE' ||
            playerState === 'BUFFERING'
        ) {
            playState = 'pause'
        }

        if (this.playState !== playState) {
            if (playState === 'play') {
                this.setPowerState('on')
                this.clearIdleTimer()
            } else {
                this.setIdleTimer()
            }
            this.playState = playState
            this.emit('playState', this.playState)
        }
    }

    setMedia(media: Media) {
        if (
            media.artist !== this.currentMedia.artist ||
            media.title !== this.currentMedia.title
        ) {
            this.currentMedia = media
            this.emit('media', this.currentMedia)
        }
    }

    updateVolume(volume: Volume) {
        this.currentVolume = volume;
        this.emit('volume', this.currentVolume)
    }

    setApplication(application: string) {
        if (application === 'Backdrop') {
            this.setPowerState('off')
            this.application = application
        } else {
            this.setPowerState('on')
            if (this.playState === 'pause') {
                this.setIdleTimer()
            }
            if (application !== this.application) {
                this.application = application
                this.emit('application', this.application)
            }
        }
    }

    setIdleTimer() {
        debug('setIdleTimer')
        this.clearIdleTimer()
        this.idleTimer = setTimeout(() => {
            debug('idleTimeout')
            this.setPowerState('off')
        }, this.idleTimeout)
    }

    clearIdleTimer() {
        debug('clearIdleTimer')
        if (this.idleTimer !== undefined) {
            debug('idleTimer cleared')
            clearTimeout(this.idleTimer)
            this.idleTimer = undefined
        }
    }

    stopDevice() {
        if (this.clientConnection) {
            this.clientConnection.stopDevice()
        }
    }

    pauseDevice() {
        if (this.clientConnection && this.playState === 'play') {
            this.clientConnection.pauseDevice()
        }
    }

    playDevice() {
        if (this.clientConnection && this.playState === 'pause') {
            this.clientConnection.playDevice()
        }
    }

    volumeUp() {
        if (this.clientConnection) {
            this.clientConnection.volumeUp()
        }
    }

    volumeDown() {
        if (this.clientConnection) {
            this.clientConnection.volumeDown()
        }
    }

    setVolume(level: number) {
        if (this.clientConnection) {
            this.clientConnection.setVolume(level)
        }
    }
}

class ClientConnection {
    private client: any
    private interval: NodeJS.Timer
    private active = false
    private reconnectTimer: NodeJS.Timer
    private receiver: any
    private transportId: string
    private sessionId: string
    private mediaConnection: MediaConnection
    private volume: Volume

    constructor(private address: string, private monitor: DeviceMonitor) {
        this.connect()
    }

    connect() {
        this.active = true
        this.client = new Client()

        this.client.on('error', (error: any) => {
            debug('error event', error)
            this.close()
            this.reconnectTimer = setTimeout(() => this.connect(), 5000)
        })

        this.client.connect(this.address, () => {
            debug('Connected')

            const connection = this.client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.tp.connection',
                'JSON',
            )
            this.receiver = this.client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.receiver',
                'JSON',
            )
            const heartbeat = this.client.createChannel(
                'sender-0',
                'receiver-0',
                'urn:x-cast:com.google.cast.heartbeat',
                'JSON',
            )

            connection.send({ type: 'CONNECT' })

            connection.on('message', (data: any) => {
                debug('connection message', data)
                if (data.type === 'CLOSE') {
                    // Never seen this happen
                    debug('Connection closed')
                    this.close()
                }
            })

            connection.on('close', () => debug('connection closed'))
            connection.on('disconnect', () => debug('connection disconnect'))

            this.interval = setInterval(() => {
                heartbeat.send({ type: 'PING' })
            }, 5000)

            this.receiver.send({
                type: 'GET_STATUS',
                requestId: requestId++,
            })

            this.receiver.on('message', (data: any) => {
                if (!data.status.applications) return
                this.monitor.setApplication(
                    data.status.applications[0].displayName,
                )

                let transportId = data.status.applications[0].transportId
                this.volume = data.status.volume.level
                debug('volume', this.volume)
                this.monitor.updateVolume(this.volume)
                if (this.transportId !== transportId) {
                    this.transportId = transportId
                    this.mediaConnection = new MediaConnection(
                        this.client,
                        this.transportId,
                        this.monitor,
                    )
                }
                this.sessionId = data.status.applications[0].sessionId
            })
        })
    }

    close() {
        if (this.active) {
            this.active = false
            clearInterval(this.interval)
            this.client.close()
            this.receiver = undefined
        } else {
            clearTimeout(this.reconnectTimer)
        }
    }

    stopDevice() {
        if (this.receiver) {
            this.receiver.send({
                type: 'STOP',
                requestId: requestId++,
                sessionId: this.sessionId,
            })
        }
    }

    pauseDevice() {
        if (this.mediaConnection) {
            this.mediaConnection.pause()
        }
    }

    playDevice() {
        if (this.mediaConnection) {
            this.mediaConnection.play()
        }
    }

    volumeUp() {
        this.receiver.send({
            type: 'SET_VOLUME',
            volume: { level: this.volume + 0.05 },
            requestId: requestId++,
        })
    }

    volumeDown() {
        this.receiver.send({
            type: 'SET_VOLUME',
            volume: { level: this.volume - 0.05 },
            requestId: requestId++,
        })
    }

    setVolume(level: number) {
        this.receiver.send({
            type: 'SET_VOLUME',
            volume: { level: level },
            requestId: requestId++,
        })
    }
}

class MediaConnection {
    private mediaSessionId: string
    private media: any

    static readonly MUSIC_TRACK = 3;

    constructor(
        client: any,
        transportId: string,
        private monitor: DeviceMonitor,
    ) {
        debug('new MediaConnection', transportId)
        let mediaConnection = client.createChannel(
            'client-17558',
            transportId,
            'urn:x-cast:com.google.cast.tp.connection',
            'JSON',
        )

        this.media = client.createChannel(
            'client-17558',
            transportId,
            'urn:x-cast:com.google.cast.media',
            'JSON',
        )

        mediaConnection.send({ type: 'CONNECT' })

        mediaConnection.on('message', (data: any) => {
            debug('mediaConnection message', data)
            if (data.type === 'CLOSE') {
                // Happens when you press Stop from the cast-dialog in Chrome
                debug('Closing MediaConnection')
                mediaConnection.close()
                this.media.close()
            }
        })

        this.media.send({
            type: 'GET_STATUS',
            requestId: requestId++,
        })

        this.media.on('message', (message: any) => this.parseMessage(message))

        this.media.on('close', () => {
          // This is needed particularly for Chromecast audio & enabled speakers
          // Update internal state, but leave the other state variables alone
          this.monitor.setPowerState('off')
        })
    }

    pause() {
        this.media.send({
            type: 'PAUSE',
            mediaSessionId: this.mediaSessionId,
            requestId: requestId++,
        })
    }

    play() {
        this.media.send({
            type: 'PLAY',
            mediaSessionId: this.mediaSessionId,
            requestId: requestId++,
        })
    }

    parseMessage(message: any) {
        debug('media message', message.requestId)
        if (message.type === 'MEDIA_STATUS') {
            let status = message.status[0]
            if (status) {
                this.mediaSessionId = status.mediaSessionId
                this.monitor.setPlayState(status.playerState)
                if (status.media && status.media.metadata) {
                    this.monitor.setMedia({
                        artist: status.media.metadata.artist,
                        title: status.media.metadata.title,
                        albumArtist: (status.media.metadata.metadataType === MediaConnection.MUSIC_TRACK) ? status.media.metadata.albumArtist : undefined,
                        albumName: (status.media.metadata.metadataType === MediaConnection.MUSIC_TRACK) ? status.media.metadata.albumName : undefined,
                        images: status.media.metadata.images,
                        trackNumber: (status.media.metadata.metadataType === MediaConnection.MUSIC_TRACK) ? status.media.metadata.trackNumber : undefined,
                    })
                }
            }
        }
    }
}
